akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -301,28 +303,25 @@ private[stream] object ActorGraphInterpreter {
 /**
  * INTERNAL API
  */
-private[stream] class ActorGraphInterpreter(
+private[stream] final class GraphInterpreterShell(
   assembly: GraphAssembly,
   inHandlers: Array[InHandler],
   outHandlers: Array[OutHandler],
   logics: Array[GraphStageLogic],
   shape: Shape,
   settings: ActorMaterializerSettings,
-  mat: Materializer) extends Actor {
+  mat: ActorMaterializerImpl) {
+
   import ActorGraphInterpreter._
 
-  val interpreter = new GraphInterpreter(
-    assembly,
-    mat,
-    Logging(this),
-    inHandlers,
-    outHandlers,
-    logics,
-    (logic, event, handler) ⇒ self ! AsyncInput(logic, event, handler),
-    settings.fuzzingMode)
+  private var self: ActorRef = _
+  lazy val log = Logging(mat.system.eventStream, self)
+
+  lazy val interpreter = new GraphInterpreter(assembly, mat, log, inHandlers, outHandlers, logics,
+    (logic, event, handler) ⇒ self ! AsyncInput(this, logic, event, handler), settings.fuzzingMode)
```

Comment:
so, by being able to compare the current shell, with the target shell in a AsyncCallback.invoke, we can remove this self message.